doc/sphinx/proof-engine/tactics.rst

```diff
@@ -36,6 +36,18 @@ language will be described in Chapter :ref:`ltac`.
 Common elements of tactics
 --------------------------
 
+Reserved keywords
+~~~~~~~~~~~~~~~~~
+
+The tactics described in this chapter reserve the following keywords::
+
+  by using
```

Comment:
These are keywords everywhere in Coq, not just in tactics.  I don't quite see the benefit of splitting up this information.  I can't imagine why users would care whether they're defined in tactics or elsewhere.  It does make sense for keywords defined in plugins that are not automatically loaded.

Same thing for the tokens mentioned below and the ltac keywords.

Reply:
> > by using
> These are keywords everywhere in Coq, not just in tactics. I don't quite see the benefit of splitting up this information. I can't imagine why users would care whether they're defined in tactics or elsewhere. It does make sense for keywords defined in plugins that are not automatically loaded.

The keywords `by` and `using` are technically used only in tactics (and unavailable in `coqtop -nois`), but they are somehow universal and that makes sense to reserve them from the beginning.

The keywords `multimatch` and `lazymatch` are on the other side pretty ad hoc (and inexistent e.g. in Ltac2). As you say, they are used in limited contexts.

Reply:
I agree `multimatch` and `lazymatch` ad hoc.  It's reasonable to consider changing them to identifiers.  But please, let's document all the keywords in one place (except those from optionally-loaded plugins).

Reply:
> But please, let's document all the keywords in one place (except those from optionally-loaded plugins).

I'm unsure of what you exactly mean. For instance, where do you want to document [=, ||, |-, **, &&, ~, etc.

Reply:
> I agree multimatch and lazymatch ad hoc. It's reasonable to consider changing them to identifiers.

Actually, they technically don't need to be keywords - adding a commit that moves them to identifiers.

(So we are back to the question of what identifiers - among those which don't technically need to be keywords from the parsing point of view - we don't want the user to use for a different purpose :smile:.)

Reply:
> I'm unsure of what you exactly mean. For instance, where do you want to document [=, ||, |-, **, &&, ~, etc.

Sorry.  I mean that documenting the fact that [=, || etc are tokens should remain in the Lexical Conventions in basic.rst.  The description of their semantics should be elsewhere.  (Note that this material was just moved to that file in merge a couple hours ago.)

Reply:
What I mean is also where to put the limit. Should be collected early all the symbols used in Ltac, in Ltac2, in the prelude, ... ? Or only a selection of them (e.g. the less ad hoc ones).

Reply:
I expect users want to see the list of keywords and tokens in order to avoid or understand clashes.  I think symbols from always-loaded plugins should be included in the front section.  For the remaining plugins, some users might be happier with a single list covering all of the optional plugins in the same section, while others may want the specifics for each plugin in the section for the plugin.  We could do both.  That places slightly more burden on those who update the doc.  Doing one upfront is the easiest--and there are plenty of other things worth improving in the doc.

When I updated this section some months ago, I didn't consider the prelude because there are so many other files that define new notations.  Also updating the list is a manual process that requires some research; very easy to missing changes.  I suppose we could mention the `Print Visibility` command up front as a way to see tokens (better than nothing), or perhaps we could create a `Print Keywords` and/or `Print Tokens` command.

Reply:
To implement a `Print Keywords` is actually the purpose of #11117.

Reply:
@jfehrle Since notations loaded in the prelude can define keywords or tokens, but we choose not to include those, I don't see why we should include in the main list the keywords and tokens defined at the level of plugins which are loaded in the prelude. From the users' point of view, this is equivalent.

However, it could make sense to say: here are the additional tokens and keywords defined in the prelude (either in notations or in plugins).

As for the plugins that are not loaded by default, it wouldn't make more sense to document their own tokens and keywords in a central place of the documentation than it would for every token and keyword ever defined by a notation in the standard library.

Reply:
In your last sentence, I'm not sure what you mean by "it wouldn't make more sense".  Should that be "wouldn't it make more sense"?

"than" suggests you're comparing 2 alternatives for where to put the info, but the wording for the second alternative doesn't really say where you want to put things (I'm guessing you're referring putting a little bit of info in each chapter.)

I think it's fine to put the list of tokens in a central location.  I don't think it's essential to distinguish the items that are loaded in the prelude.

Reply:
Yes I am comparing two alternatives, one of which is completely unrealistic, to argue against the other one.

I am saying the following: since it is unrealistic to document every keyword / token ever defined by a notation in the standard library in a central place of the reference manual, it is not justified to document all the keywords / tokens defined by plugins that are not loaded by default in a central place. Indeed, we should not make a difference between keywords / tokens defined in OCaml or Coq code.

>I don't think it's essential to distinguish the items that are loaded in the prelude.

I am not sure here if you are talking of distinguishing these items from the ones that are defined even if you don't load the prelude, or from the ones that are defined in plugins that are not loaded by default. To the second, I think I've just answered. To the first, I still think it is useful to make the distinction because using Coq with the `-noinit` flag is definitely a use case that we do support.

Reply:
> it is unrealistic to document every keyword / token ever defined by a notation in the standard library in a central place of the reference manual, it is not justified to document all the keywords / tokens defined by plugins that are not loaded by default in a central place

That's fine.

>I am not sure here if you are talking of distinguishing these items from the ones that are defined even if you don't load the prelude, or from the ones that are defined in plugins that are not loaded by default. To the second, I think I've just answered. To the first, I still think it is useful to make the distinction because using Coq with the -noinit flag is definitely a use case that we do support.

Also fine.  I meant distinguishing the -noinit case from the regular startup case, neither of which include items from optional plugins.  (Figuring out the set of all keywords/tokens defined in all or them would be pretty tedious!)

I thought @herbelin was eager to include the info for optional plugins.
In https://github.com/coq/coq/pull/12229#discussion_r419024267 I was merely discussing 2 ways to present that info--I don't think it's essential to include.